File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
        public int[] productExceptSelf(int[] nums) {
        int arrayLen = nums.length;
        int[] ans_array = new int[arrayLen];
        Arrays.fill(ans_array, 1);

        int prefixValue = 1;
        int postfixValue = 1;

        for (int ind = 0; ind < arrayLen; ind++) {
            ans_array[ind] = prefixValue;
            prefixValue *= nums[ind];
        }

        for (int postInd = arrayLen-1; postInd > -1; postInd--) {
            ans_array[postInd] = ans_array[postInd] * postfixValue;
            postfixValue = nums[postInd] * postfixValue;
        }

        return ans_array;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        array_len = len(nums)
        ans_arr = [0] * array_len
        prefix_product = [1] * array_len
        prefix_product[0] = nums[0]

        postfix_product = [1] * array_len
        postfix_product[-1] = nums[-1]

        for ind in range(array_len):
            if ind == 0:
                prefix_product[ind] = 1
            else:
                prefix_product[ind] = prefix_product[ind-1] * nums[ind-1]

        for ind in range(array_len-1, -1, -1):
            if ind == array_len - 1:
                postfix_product[ind] = 1
            else:
                postfix_product[ind] = postfix_product[ind+1] * nums[ind+1]

        for final_ind in range(array_len):
            ans_arr[final_ind] = prefix_product[final_ind] * postfix_product[final_ind]
        
        return ans_arr